kv/etcd/etcd.go

```diff
@@ -0,0 +1,292 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/protobuf/proto"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"github.com/uber-go/tally"
+	"golang.org/x/net/context"
+)
+
+var noopCancel func()
+
+// NewStore creates a kv store based on etcd
+func NewStore(etcd *clientv3.Client, opts Options) kv.Store {
+	scope := opts.InstrumentsOptions().MetricsScope()
+	return &client{
+		opts:       opts,
+		kv:         etcd.KV,
+		watcher:    etcd.Watcher,
+		watchables: map[string]kv.ValueWatchable{},
+		retrier:    xretry.NewRetrier(opts.RetryOptions()),
+		logger:     opts.InstrumentsOptions().Logger(),
+		m: clientMetrics{
+			etcdGetError:    scope.Counter("etcd-get-error"),
+			etcdPutError:    scope.Counter("etcd-put-error"),
+			etcdTnxError:    scope.Counter("etcd-tnx-error"),
+			etcdWatchCreate: scope.Counter("etcd-watch-create"),
+			etcdWatchError:  scope.Counter("etcd-watch-error"),
+			etcdWatchReset:  scope.Counter("etcd-watch-reset"),
+		},
+	}
+}
+
+type client struct {
+	sync.RWMutex
+
+	opts       Options
+	kv         clientv3.KV
+	watcher    clientv3.Watcher
+	watchables map[string]kv.ValueWatchable
+	retrier    xretry.Retrier
+	logger     xlog.Logger
+	m          clientMetrics
+}
+
+type clientMetrics struct {
+	etcdGetError    tally.Counter
+	etcdPutError    tally.Counter
+	etcdTnxError    tally.Counter
+	etcdWatchCreate tally.Counter
+	etcdWatchError  tally.Counter
+	etcdWatchReset  tally.Counter
+}
+
+func (c *client) Get(key string) (kv.Value, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	r, err := c.kv.Get(ctx, c.opts.KeyFn()(key))
+	if err != nil {
+		c.m.etcdGetError.Inc(1)
+		return nil, err
+	}
+
+	if r.Count == 0 {
+		return nil, kv.ErrNotFound
+	}
+
+	if r.Count > 1 {
+		return nil, fmt.Errorf("received %d values for key %s, expecting 1", r.Count, key)
+	}
+
+	return kv.NewValue(r.Kvs[0].Value, int(r.Kvs[0].Version)), nil
+}
+
+func (c *client) Watch(key string) (kv.ValueWatch, error) {
+	c.Lock()
+	watchable, ok := c.watchables[key]
+	if !ok {
+		watchChan := c.watcher.Watch(
+			context.Background(),
+			c.opts.KeyFn()(key),
+			clientv3.WithProgressNotify(),
```

Comment:
This seems somewhat hackish, basically we are using WithProgressNotify to bound the maximum staleness of the keys on the client side, which is not quite reliable (e.g., server may determine when to send depending on the server load etc). If we are worried that `Get` could fail even with retries, can we simply increase the number of retries with an upper bound on the maximum backoff delay (e.g., 10 minutes)?